### ISSUE TITLE
PoR: Tinker-like Megatron TrainingEngine APIs

### ISSUE THREAD

**Issue by yaoyu-33** [NVIDIA-NeMo/Megatron-Bridge#4515]
## Goal

Design and stage a Tinker-like Megatron API layer for Megatron Bridge.

This tracker focuses on the user-facing idea and UX, not the detailed implementation plan. The goal is to make Megatron easier to adopt for post-training and custom algorithm loops while preserving control over model, loss, rollout, checkpoint, and export semantics.

## Background

Today users generally choose between two adoption paths:

1. Run Megatron-LM directly through its training stack. This gives maximum control over model, parallelism, optimizer, data, and training-loop details, but requires users to understand and often modify Megatron internals.
2. Run Megatron Bridge recipes. Recipes provide a much easier starting point for OSS models, especially pretraining and fine-tuning, but training-loop-level customization is not straightforward.

The current experience is close to "all or nothing": users either work directly in Megatron internals, or use a recipe that hides most of the training loop.

We want a middle layer: simple, flexible APIs that let users keep algorithmic control while Bridge handles the repeated distributed-training mechanics.

## Motivation From Existing Integrations

We investigated training libraries that integrate with Megatron-style backends, including verl, miles, SkyRL, and NeMo RL. These systems often use Megatron-LM and/or Megatron Bridge primitives, but avoid the Bridge `pretrain` API because they need finer-grained control.

Common repeated work across these frameworks includes:

- deriving Megatron providers from HF models, then overriding parallelism, dtype, packing, MoE, PEFT, DDP/FSDP, and other runtime choices
- deciding valid and efficient recipe settings outside Bridge
- registering pre-wrap hooks for PEFT, value heads, frozen routers, local model patches, and custom role behavior
- moving model state between GPU and CPU for colocated training / rollout workloads
- managing multiple model roles such as actor, reference, critic/value, and reward models
- converting framework batches into Megatron-compatible tensors, masks, padding, packing, `PackedSeqParams`, CP/TP alignment, and microbatches
- adapting custom losses to Megatron forward/backward schedule contracts
- normalizing losses by valid response tokens or samples while compensating for Megatron schedule, CP, microbatch, and DP scaling behavior
- saving/loading training state across TP/PP/VPP/CP/EP sharding, distributed optimizer state, PEFT adapters, FSDP-DTensor, and HF export formats

This overlap suggests these mechanics should be centralized in Megatron Bridge rather than repeated by each downstream framework.

## Proposed UX Direction

Inspired by Tinker-style APIs, Bridge should expose a clearer responsibility split:

- Infrastructure handles distributed Megatron model construction, parallelism, wrapping, packing/padding, schedule selection, output restoration, loss scaling, optimizer stepping, checkpointing, and export.
- Users own the algorithm loop: rollout, reward logic, loss choice, optimizer LR schedule, when to forward/backward, when to step, when to save, and when to export.

The main proposed abstraction is `TrainingEngine`.

```python
engine = megatron.bridge.TrainingEngine(
    model="Qwen/Qwen3-8B",
    checkpoint_path=None,
    output_head="lm",
    trainable=True,
    lora=megatron.bridge.LoraConfig(rank=32),
    hooks=None,
    overrides={"tensor_model_parallel_size": 4},
)
```

`TrainingEngine` is an in-process distributed API created on every rank inside the launched GPU job. It is not initially a CPU-side service client.

The recipe is the starting point, not a black box. Bridge resolves a recommended baseline recipe for the selected model, merges explicit overrides, initializes Megatron parallelism, builds/wraps the model, loads pretrained or resume checkpoints, creates optimizer/scheduler state when needed, and installs hooks before wrapping.

### Model Movement

```python
engine.cuda()
engine.cpu()
```

These APIs should hide movement of parameters, gradients, DDP/FSDP buffers, expert buffers, frozen params, and optimizer state.

### Multiple Model Roles

Users can instantiate separate engines for actor, critic/value, reference, and reward models through explicit configuration rather than hard-coded role semantics.

```python
actor = TrainingEngine(model="Qwen/Qwen3-8B", output_head="lm", trainable=True, lora=LoraConfig(rank=32))
critic = TrainingEngine(model="Qwen/Qwen3-8B", output_head="value", trainable=True)
ref = TrainingEngine(model="Qwen/Qwen3-8B", output_head="lm", trainable=False)
```

For LoRA training, users should also be able to run a base-model forward by disabling adapters:

```python
ref_outputs = actor.forward(datums, disable_adapters=True)
```

### Forward / Backward

```python
output = engine.forward(datums)
train_output = engine.forward_backward(datums, loss_fn="importance_sampling")
last_stage_output = engine.forward(datums, broadcast_output=False)
```

The engine should choose the Megatron schedule internally based on PP/VPP configuration and adapt the user batch/loss contract to the schedule contract.

### Datum Contract

`Datum` is the user/Megatron boundary. Users provide logical samples; the engine handles Megatron-specific packing, padding, `PackedSeqParams`, CP/TP/PP layout, microbatching, output restoration, and loss scaling.

```python
datum = megatron.bridge.Datum(
    input_ids=input_ids,
    loss_inputs={
        "target_tokens": target_tokens,
        "logprobs": rollout_logprobs,
        "advantages": advantages,
        "weights": loss_mask,
    },
)
```

TrainingEngine should expose only data-parallel sharding to users:

```python
engine.data_rank
engine.data_world_size
```

TP/CP/EP/PP layout should be internal to the engine for entry-level users.

### ModelOutput Contract

Outputs should be reconstructed into per-datum semantics for the local data-parallel rank.

```python
@dataclass
class ModelOutput:
    loss: torch.Tensor | None = None
    logprobs: list[torch.Tensor] | None = None
    entropy: list[torch.Tensor] | None = None
    values: list[torch.Tensor] | None = None
    metrics: dict[str, float] = field(default_factory=dict)
```

With `broadcast_output=True`, all PP ranks receive populated outputs. With `False`, only the output PP stage returns populated tensor fields.

### Built-in and Custom Losses

Built-ins should cover common post-training objectives over time, starting with cross entropy and moving toward importance sampling / PPO-style losses.

Custom losses should operate over `ModelOutput` and the original datums:

```python
LossFn = Callable[[ModelOutput, Sequence[Datum]], Sequence[torch.Tensor]]
```

TrainingEngine handles valid-token weighting, global denominators across data ranks, Megatron microbatch scaling, CP compensation, backward, gradient reduction, and optimizer state.

### Optimizer Step

```python
metrics = engine.optim_step(lr=lr)
```

The optimizer config is still recipe/override driven, but the algorithm loop owns the per-step LR.

### Checkpoint

Checkpoint APIs are for training resume, not HF-compatible export.

```python
handle = engine.save_state(path, user_state={"step": step}, async_save=True)
handle.wait()
user_state = engine.load_state(path)
```

The checkpoint should include Megatron training state: model shards, optimizer, scheduler, RNG, recipe/override metadata, and optional user state.

### Export

Export APIs are for HF-compatible weights or inference-engine weight sync, not training resume.

```python
engine.export(path="exports/step_100")
for name, tensor in engine.export(streaming=True):
    inference_engine.update_weight(name, tensor)
```

Streaming export must remain a distributed collective; all ranks enter and consume the generator.

## Stage 1 Scope

This issue tracks Stage 1. Stage 1 should focus on local in-process APIs before a service/client runtime.

- [ ] Define the public `TrainingEngine`, `Datum`, and `ModelOutput` contracts.
- [ ] Prototype recipe resolution + override merging for a small set of supported models.
- [ ] Improve model builder ergonomics around provider construction and hook installation.
- [ ] Add high-level pre-wrap hook policies for common cases: LoRA, value head, frozen parts, local patches.
- [ ] Add model role examples: actor, reference, critic/value.
- [ ] Add `forward()` for inference/logprob/value extraction with adapter disabling support.
- [ ] Add `forward_backward()` with built-in `cross_entropy` first.
- [ ] Define the loss scaling contract for token-level and sample-level losses.
- [ ] Add `optim_step(lr=...)` wrapper semantics.
- [ ] Define and prototype coarse `engine.cpu()` / `engine.cuda()` movement APIs.
- [ ] Study existing offload/reload behavior in downstream frameworks and map required state: model, optimizer, buffers, gradients, expert buffers, distributed optimizer state.
- [ ] Define checkpoint save/load policy for training resume.
- [ ] Define export policy for HF-compatible export and streaming inference weight sync.
- [ ] Add LoRA support in the engine path.
- [ ] Track Multi-LoRA requirements separately or include a first minimal design if needed.

## Out of Scope For Stage 1

- A CPU-side service/client runtime.
- Scheduler or cluster resource allocation abstractions.
- Full Tinker SDK compatibility.
- A production sampling service.
- Complete PPO/GRPO implementation.

These are important long-term directions, but Stage 1 should first prove the in-process `TrainingEngine` contract.

## Long-Term Direction

A later service-like UX could expose a CPU-side client similar to Tinker:

1. Launch a service runtime on the cluster.
2. Connect from a CPU-side process.
3. Create training clients backed by service-managed Megatron workers.
4. Reuse Megatron inference serving for rollout sampling where possible.
5. Support weight sync from `TrainingEngine.export(streaming=True)` into an inference runtime.

There is also an open question of whether Megatron should implement a Tinker-compatible backend contract so users can keep a Tinker-style SDK while the backend executes through Megatron.

## Open Questions

1. Should `TrainingEngine` live directly in Megatron Bridge, or in a separate package that depends on Bridge?
2. What is the minimum `Datum` contract that covers SFT, DPO, PPO/GRPO, value training, and multimodal batches without becoming too broad?
3. Which outputs should `ModelOutput` expose initially: logits, logprobs, entropy, values, hidden states, or selected-token-only tensors?
4. How should custom loss normalization be made explicit while still protecting users from Megatron microbatch / CP / DP scaling pitfalls?
5. How much automatic recipe scaling is safe when users launch with more GPUs than a baseline recipe needs?
6. What should be built into Stage 1 versus left to downstream frameworks?
7. What is the right relationship with Megatron inference APIs for rollout sampling and weight refit?
8. Is Tinker-compatible protocol support strategically worth pursuing?

## References To Attach

- verl Megatron integration investigation
- miles Megatron integration investigation
- SkyRL Megatron integration investigation
- NeMo RL Megatron integration investigation
- Tinker API / cookbook examples
